src/Query.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL;
+
+use Doctrine\DBAL\Types\Type;
+
+/**
+ * A SQL query together with its bound parameters.
+ *
+ * @psalm-immutable
+ */
+class Query
+{
+    /**
+     * The SQL query.
+     *
+     * @var string
+     */
+    private $sql;
+
+    /**
+     * The parameters bound to the query, if the exception occurred during a query.
+     *
+     * @var array<mixed>
+     */
+    private $params;
+
+    /**
+     * The types of the parameters bound to the query, if the exception occurred during a query.
+     *
+     * @var array<Type|int|string|null>
+     */
+    private $types;
+
+    /**
+     * @param array<mixed>                $params
+     * @param array<Type|int|string|null> $types
+     *
+     * @psalm-suppress ImpurePropertyAssignment
```

Comment:
I haven't found a way around this: Psalm is right that `Type` is not immutable.

Reply:
I see. Makes sense.